Add missing jo options

hartzell opened the pull request:
Add support for several options that jo supports but gjo does not

- -B -- disable treating true/false as bool
- -e -- empty stdin is not an error
- -n -- ignore keys with empty values
- read from stdin if not args are supplied

It adds code to handle the various features, modifies the run function
to take an io.Reader as an argument for testability, and calls the run
function with os.Stdin in main.

It modifies the TestRun test:

- The test cases have been updated:
    - renamed fields to be clear about meaning
    - only non-zero test values need to be supplied in the test case
    - provides stdin as a string.Reader
    - provides want_stdout to hold the expected output
- Tests have been added for new features

Commit:
Ignore the gjo executable via .gitignore

Don't risk committing the result of `go build`, ignore it via
.gitignore.

Commit:
Add support for missing jo features

Add support for several options that jo supports but gjo does not

- -B -- disable treating true/false as bool
- -e -- empty stdin is not an error
- -n -- ignore keys with empty values
- read from stdin if not args are supplied

It adds code to handle the various features, modifies the run function
to take an io.Reader as an argument for testability, and calls the run
function with os.Stdin in main.

It modifies the TestRun test:

- The test cases have been updated:
    - renamed fields to be clear about meaning
    - only non-zero test values need to be supplied in the test case
    - provides stdin as a string.Reader
    - provides want_stdout to hold the expected output
- Tests have been added for new features

skanehira commented on main.go:
```diff
@@ -20,9 +21,12 @@ type Version struct {
-	pretty  = flag.Bool("p", false, "pretty-prints")
-	version = flag.Bool("v", false, "show version")
+	array              = flag.Bool("a", false, "creates an array of words")
+	disable_bool       = flag.Bool("B", false, "disable treating true/false as bool")
+	ignore_empty_stdin = flag.Bool("e", false, "empty stdin is not an error")
+	ignore_empty_keys  = flag.Bool("n", false, "ignore keys with empty values")
```
We usually use camelcase in Go.

hartzell replied:
done, will push shortly.

skanehira commented on main_test.go:
```diff
@@ -110,35 +111,100 @@ func TestVersion(t *testing.T) {
-		for arg, value := range test.args {
-			flag.CommandLine.Set(arg, value)
+		os.Args = append([]string{""}, test.args...)
+		for test_flag, value := range _flags(test.flags) {
+			flag.CommandLine.Set(test_flag, value)
 		}
```
```suggestion
		for flag, value := range _flags(test.flags) {
			flag.CommandLine.Set(flag, value)
		}
```

hartzell replied:
Can't use flag, the LSP output sums it up nicely:

> renaming this var "test_flag" to "flag"	would shadow this reference	to the imported package name declared here

skanehira commented on main_test.go:
```diff
@@ -110,35 +111,100 @@ func TestVersion(t *testing.T) {
+
+// _flags merges a set of override values into the standard set of
+// command line flags, overriding the default values.
+func _flags(overrides map[string]string) map[string]string {
```
```suggestion
func defaultFlags(overrides map[string]string) map[string]string {
```

hartzell replied:
done, will push shortly

skanehira commented on main.go:
```diff
@@ -133,7 +149,7 @@ func doVersion() error {
 }
 
-func run() int {
+func run(stdin io.Reader) int {
```
For testability, I think we should accept `args` and return `result`.

```go
func run(args []string) string
```

hartzell replied:
Happy to make changes here, but I don't quite understand what you want this to look like.  Can you explain more fully?

1. I think you're suggesting to pass in the `args` instead of referring to the global `flag.Args()`.  Are you suggesting parsing the args and etc.. in `main`?
2. I find passing in an io.Reader instead of relying on os.Stdin or mocking it or using real files makes the table driven tests in main_test.go:TestRun cleaner (possible). 
3. What result do you want to return?  It's currently returning success/failure, which is easily testable in `main_test.go:TestRun`, along with the output and stderr.

hartzell committed:
Implement code review requests

- use camelCase instead of snake_case for variable names (sorry, I've
  been dealing with a bunch of python...).
- rename functiont that returns default flags to defaultFlags.
